tron/core/actionrun.py

```diff
@@ -572,15 +591,43 @@ def cleanup(self):
         self.clear_observers()
         if self.triggered_by:
             EventBus.clear_subscriptions(self.__hash__())
+        self.clear_trigger_timeout()
         self.cancel()
 
+    def clear_trigger_timeout(self):
+        if self.trigger_timeout_call:
+            self.trigger_timeout_call.cancel()
+            self.trigger_timeout_call = None
+
     def setup_subscriptions(self):
-        for trigger_pattern in self.triggered_by or []:
-            trigger = self.render_template(trigger_pattern)
-            EventBus.subscribe(trigger, self.__hash__(), self.trigger_notify)
+        if self.triggered_by:
+            if self.remaining_triggers:
+                now = timeutils.current_time().timestamp()
+                delay = max(self.trigger_timeout_timestamp - now, 1)
+                self.trigger_timeout_call = reactor.callLater(
+                    delay, self.trigger_timeout_reached
+                )
+
+            for trigger_pattern in self.triggered_by:
+                trigger = self.render_template(trigger_pattern)
+                if not EventBus.has_event(trigger):
+                    EventBus.subscribe(
+                        trigger, self.__hash__(), self.trigger_notify
+                    )
+
+    def trigger_timeout_reached(self):
+        if self.remaining_triggers:
+            self.trigger_timeout_call = None
+            log.warning(
+                f"{self} reached timeout waiting for: {self.remaining_triggers}"
+            )
+            self.fail(self.EXIT_TRIGGER_TIMEOUT)
+        else:
```

Comment:
This shouldn't happen normally, right? (the timeout being called but the triggers have finished)

Reply:
it can happen, just very unlikely. decided to put the code in for completeness anyway.